### PR TITLE
Bootstrapping bootstrap checks

### LINKED ISSUE #17474: Production startup checks should output all problems at once
The production start up checks (e.g. minimum master nodes, max file descriptors) should check all the settings and output a single error containing all the problems rather than showing the first error and then waiting for it to be fixed and the node started again before showing the next error.

### PR DESCRIPTION
**Pull request by jasontedor**
This pull request offers two enhancements to the bootstrap checks.

The first is to execute the bootstrap checks in a way that we can detect
whether or not we our bound to a non-loopback interface or publishing to
a non-loopback interface (instead of merely checking whether or not any
of the network settings are set but not what their value is). This is
accomplished by moving the execution of the bootstrap checks to after
the networking services are started; as a bonanza, this enables us to
output the node name with the failing checks, a capability that did not
exist before when these messages were logged before the node had even
started.

The second is a modification to the bootstrap checks to output all
failing checks instead of forcing the user to resolve them one-by-one.

Closes #17474, #17570

### CODE REVIEW

**Comment by jasontedor** [comment]
Now we have this:

```bash
$ ~/elasticsearch/elasticsearch-5.0.0-alpha1-SNAPSHOT/bin/elasticsearch -E es.network.host=192.168.1.8
[2016-04-07 10:02:46,394][INFO ][node                     ] [Texas Twister] version[5.0.0-alpha1-SNAPSHOT], pid[37835], build[e60cf58/2016-04-07T13:34:52.575Z]
[2016-04-07 10:02:46,395][INFO ][node                     ] [Texas Twister] initializing ...
[2016-04-07 10:02:46,664][INFO ][plugins                  ] [Texas Twister] modules [lang-mustache, lang-painless, ingest-grok, reindex, lang-expression, lang-groovy], plugins []
[2016-04-07 10:02:46,686][INFO ][env                      ] [Texas Twister] using [1] data paths, mounts [[/ (/dev/disk1)]], net usable_space [109.8gb], net total_space [232.6gb], spins? [unknown], types [hfs]
[2016-04-07 10:02:46,686][INFO ][env                      ] [Texas Twister] heap size [989.8mb], compressed ordinary object pointers [true]
[2016-04-07 10:02:48,245][INFO ][node                     ] [Texas Twister] initialized
[2016-04-07 10:02:48,248][INFO ][node                     ] [Texas Twister] starting ...
[2016-04-07 10:02:48,342][INFO ][transport                ] [Texas Twister] publish_address {192.168.1.8:9300}, bound_addresses {192.168.1.8:9300}
Exception in thread "main" java.lang.RuntimeException: bootstrap checks failed
max size virtual memory [35184372088832] for user [jason] likely too low, increase to [unlimited]
please set [discovery.zen.minimum_master_nodes] to a majority of the number of master eligible nodes in your cluster.
        at org.elasticsearch.bootstrap.BootstrapCheck.check(BootstrapCheck.java:96)
        at org.elasticsearch.bootstrap.BootstrapCheck.check(BootstrapCheck.java:63)
        at org.elasticsearch.node.Node.start(Node.java:333)
        at org.elasticsearch.bootstrap.Bootstrap.start(Bootstrap.java:200)
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:263)
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:111)
        at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:106)
        at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:88)
        at org.elasticsearch.cli.Command.main(Command.java:53)
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:74)
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:67)
Refer to the log for complete error details.
[2016-04-07 10:02:48,399][INFO ][node                     ] [Texas Twister] stopping ...
[2016-04-07 10:02:48,410][INFO ][node                     ] [Texas Twister] stopped
[2016-04-07 10:02:48,411][INFO ][node                     ] [Texas Twister] closing ...
[2016-04-07 10:02:48,421][INFO ][node                     ] [Texas Twister] closed
```

where I have two failing checks with some additional detail in the logs

```
[2016-04-07 10:02:48,346][ERROR][bootstrap                ] [Texas Twister] Exception
java.lang.RuntimeException: bootstrap checks failed
max size virtual memory [35184372088832] for user [jason] likely too low, increase to [unlimited]
please set [discovery.zen.minimum_master_nodes] to a majority of the number of master eligible nodes in your cluster.
        at org.elasticsearch.bootstrap.BootstrapCheck.check(BootstrapCheck.java:96)
        at org.elasticsearch.bootstrap.BootstrapCheck.check(BootstrapCheck.java:63)
        at org.elasticsearch.node.Node.start(Node.java:333)
        at org.elasticsearch.bootstrap.Bootstrap.start(Bootstrap.java:200)
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:263)
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:111)
        at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:106)
        at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:88)
        at org.elasticsearch.cli.Command.main(Command.java:53)
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:74)
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:67)
        Suppressed: java.lang.IllegalStateException: max size virtual memory [35184372088832] for user [jason] likely too low, increase to [unlimited]
                at org.elasticsearch.bootstrap.BootstrapCheck.check(BootstrapCheck.java:85)
                ... 10 more
        Suppressed: java.lang.IllegalStateException: please set [discovery.zen.minimum_master_nodes] to a majority of the number of master eligible nodes in your cluster.
                at org.elasticsearch.bootstrap.BootstrapCheck.check(BootstrapCheck.java:85)
                ... 10 more
```

Note that we no longer fail if we set `es.network.host` to a loopback address:

```bash
± ~/elasticsearch/elasticsearch-5.0.0-alpha1-SNAPSHOT/bin/elasticsearch -E es.network.host=_local_
[2016-04-07 10:04:32,032][INFO ][node                     ] [Forgotten One] version[5.0.0-alpha1-SNAPSHOT], pid[38016], build[e60cf58/2016-04-07T13:34:52.575Z]
[2016-04-07 10:04:32,033][INFO ][node                     ] [Forgotten One] initializing ...
[2016-04-07 10:04:32,303][INFO ][plugins                  ] [Forgotten One] modules [lang-mustache, lang-painless, ingest-grok, reindex, lang-expression, lang-groovy], plugins []
[2016-04-07 10:04:32,327][INFO ][env                      ] [Forgotten One] using [1] data paths, mounts [[/ (/dev/disk1)]], net usable_space [109.8gb], net total_space [232.6gb], spins? [unknown], types [hfs]
[2016-04-07 10:04:32,328][INFO ][env                      ] [Forgotten One] heap size [989.8mb], compressed ordinary object pointers [true]
[2016-04-07 10:04:33,867][INFO ][node                     ] [Forgotten One] initialized
[2016-04-07 10:04:33,868][INFO ][node                     ] [Forgotten One] starting ...
[2016-04-07 10:04:33,944][INFO ][transport                ] [Forgotten One] publish_address {127.0.0.1:9300}, bound_addresses {[fe80::1]:9300}, {[::1]:9300}, {127.0.0.1:9300}
[2016-04-07 10:04:33,945][WARN ][bootstrap                ] [Forgotten One] max size virtual memory [35184372088832] for user [jason] likely too low, increase to [unlimited]
[2016-04-07 10:04:33,945][WARN ][bootstrap                ] [Forgotten One] please set [discovery.zen.minimum_master_nodes] to a majority of the number of master eligible nodes in your cluster.
[2016-04-07 10:04:36,984][INFO ][cluster.service          ] [Forgotten One] new_master {Forgotten One}{eg0mPFkAQ7qrAGGNKy4-Ow}{127.0.0.1}{127.0.0.1:9300}, reason: zen-disco-join(elected_as_master, [0] joins received)
[2016-04-07 10:04:37,004][INFO ][http                     ] [Forgotten One] publish_address {127.0.0.1:9200}, bound_addresses {[fe80::1]:9200}, {[::1]:9200}, {127.0.0.1:9200}
[2016-04-07 10:04:37,004][INFO ][node                     ] [Forgotten One] started
[2016-04-07 10:04:37,011][INFO ][gateway                  ] [Forgotten One] recovered [0] indices into cluster_state
```

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 1]
```diff
@@ -67,50 +66,45 @@ public static void check(final Settings settings) {
...
+        final ESLogger logger = Loggers.getLogger(BootstrapCheck.class, nodeName);
+
+        final List<IllegalStateException> exceptions = new ArrayList<>();
 
         for (final Check check : checks) {
             final boolean fail = check.check();
             if (fail) {
                 if (enforceLimits) {
-                    throw new RuntimeException(check.errorMessage());
+                    exceptions.add(new IllegalStateException(check.errorMessage()));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: Nit: I think it will be cleaner to collect messages and in the end either log them all or suppress them all and rethrow

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 2]
```diff
@@ -67,50 +66,45 @@ public static void check(final Settings settings) {
...
      *
-     * @param settings the current node settings
+     * @param boundTransportAddress the node network bindings
      * @return true if the checks should be enforced
      */
     // visible for testing
-    static boolean enforceLimits(final Settings settings) {
-        return enforceSettings().stream().anyMatch(s -> s.exists(settings));
+    static boolean enforceLimits(BoundTransportAddress boundTransportAddress) {
+        return !(Arrays.stream(boundTransportAddress.boundAddresses()).allMatch(TransportAddress::isLocalAddress) &&
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: :wave: to simon :)

File: core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java [hunk 3]
```diff
@@ -46,5 +46,9 @@
 
     boolean sameHost(TransportAddress other);
 
+    default boolean isLocalAddress() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java]: we should we have a default false here? feels strange. Also it feels that isLocalAddress might be confusing since we have the `LocalTransportAddress` class. Maybe flip it and call it isPublicAddress?

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 4]
```diff
@@ -268,6 +269,20 @@ public Environment getEnvironment() {
...
+     * Start the node but invoke the registered callback after network
+     * services are started but before the cluster service is started
+     * and before the network service starts accepting incoming network
+     * requests. If the node is already started, this method is a
+     * no-op.
+     *
+     * @param nodeStartCallback the registered callback
+     * @return the started node
+     */
+    public Node start(BiConsumer<Settings, BoundTransportAddress> nodeStartCallback) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I'm not sure about this call back.. feels hacky to pass in such a callback which is fairly obscure as to what it needs to do.

Maye have a method called `validateNodeBeforeAcceptingRequests` which does nothing by default. The bootstrap class can instantiate a new `ValidatedNode` class inheriting from Node (ala what we do in the tests - see `MockNode`) and that class will override the above method to do the checks?

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java [hunk 5]
```diff
@@ -20,26 +20,113 @@
...
+            final TransportAddress randomTransportAddress = mock(TransportAddress.class);
+            when(randomTransportAddress.isLocalAddress()).thenReturn(randomBoolean());
+            transportAddresses.add(randomTransportAddress);
+        }
+
+        final TransportAddress publishAddress = mock(TransportAddress.class);
+        when(publishAddress.isLocalAddress()).thenReturn(randomBoolean());
+
+        final BoundTransportAddress boundTransportAddress = mock(BoundTransportAddress.class);
+        when(boundTransportAddress.boundAddresses()).thenReturn(transportAddresses.toArray(new TransportAddress[0]));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: can we shuffle the address list? now the first is always public

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java [hunk 6]
```diff
@@ -20,26 +20,113 @@
...
 
 public class BootstrapCheckTests extends ESTestCase {
 
     public void testNonProductionMode() {
         // nothing should happen since we are in non-production mode
-        BootstrapCheck.check(Settings.EMPTY);
+        TransportAddress localTransportAddress = mock(TransportAddress.class);
+        when(localTransportAddress.isLocalAddress()).thenReturn(true);
+        BoundTransportAddress boundTransportAddress = mock(BoundTransportAddress.class);
+        when(boundTransportAddress.boundAddresses()).thenReturn(new TransportAddress[] { localTransportAddress });
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: strictly speaking, you can have more than one - do we care?

**Comment by bleskes** [comment]
I like the change. Left some minor suggestions

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 4 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I like this. I pushed fee9c982f89e8102c5877f612b0393fc156e39b8.

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java (hunk 5 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: I pushed b4b76642ac8cb7d67226746d423f46f230d1f1c6.

File: core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java]: I do not see the issue with the default implementation? I did rename the method though in 6f138460172805fb0665c5ccba85846151992be4.

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: I pushed f1c8b630ef9045e673917962a226fc8ea7ea802d. Don't hate. 😄

File: core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java]: well, I'm not sure we can assume that all addresses are by default "public". I tend towards saying implementers need to make this call.

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java (hunk 6 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: I pushed d42a2967693c567b2637ac48dd968f5e91732f59.

File: core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java]: Part of living in a modern Java 8 world is that when you implement an interface you also need to inspect the default methods on the interface?

**Comment by jasontedor** [comment]
@bleskes I've responded to your feedback.

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 7]
```diff
@@ -268,10 +268,6 @@ public Environment getEnvironment() {
      * Start the node. If the node is already started, this method is no-op.
      */
     public Node start() {
-        if (!lifecycle.moveToStarted()) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/node/Node.java]: why is this removed?

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 8]
```diff
@@ -67,50 +67,42 @@ public static void check(final Settings settings) {
...
+        if (!exceptions.isEmpty()) {
+            final List<String> messages = new ArrayList<>(1 + exceptions.size());
+            messages.add("bootstrap checks failed");
+            exceptions.forEach(e -> messages.add(e.getMessage()));
+            if (enforceLimits) {
+                final RuntimeException re = new RuntimeException(String.join("\n", messages));
+                exceptions.forEach(re::addSuppressed);
+                throw re;
+            } else {
+                messages.forEach(message -> logger.warn(message));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: This is what I meant, feel free to change or keep as is:

```
        final List<String> errors = checks.stream()
            .filter(BootstrapCheck.Check::check)
            .map(BootstrapCheck.Check::errorMessage)
            .collect(Collectors.toList());

        if (!errors.isEmpty()) {
            final List<String> messages = new ArrayList<>(1 + errors.size());
            messages.add("bootstrap checks failed");
            if (enforceLimits) {
                final RuntimeException re = new RuntimeException(String.join("\n", messages));
                errors.forEach(e -> re.addSuppressed(new IllegalStateException(e)));
                throw re;
            } else {
                messages.forEach(message -> logger.warn(message));
            }
        }
```

**Comment by bleskes** [comment]
LGTM with one minor exception and one suggestion - now it's illegal to call node.start() twice. I'm not saying that's good, but I don't think we should change it here.

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 7 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/node/Node.java]: That was a mistake when I collapsed the two `start` methods back down to one. I will fix, thank you for noticing!

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I pushed ddfa3a661510f25c2ce431dfd6fb86ac11eb8888.

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java (hunk 8 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: I knew what you meant and I don't see it as cleaner, but you forgot to add the `errors` to `messages`. Either way, I pushed a63dcb0ef306a7ff4dbfa8ffd41e6400bccdee77.

File: core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/transport/TransportAddress.java]: I pushed 415967989f0ba2d2f37e199cbd009e9282e7be72.
